### PR TITLE
Reading in time series

### LINKED ISSUE #219: Provide facilities to read in time series

### PR DESCRIPTION
**Pull request by ckittl**
Closes #219

### CODE REVIEW

File: src/main/java/edu/ie3/datamodel/io/source/csv/CsvTimeSeriesSource.java [hunk 1]
```diff
@@ -30,8 +66,181 @@ public CsvTimeSeriesSource(
...
                     }))
         .collect(Collectors.toSet());
   }
+
+  /**
+   * Acquire all available time series
+   *
+   * @return A container with all relevant time series
+   */
+  public TimeSeriesContainer getTimeSeries() {
```
Comment (johanneshiry) [inline src/main/java/edu/ie3/datamodel/io/source/csv/CsvTimeSeriesSource.java:79]: Pls add this to the `TimeSeriesSource` interface

File: src/main/java/edu/ie3/datamodel/io/source/csv/CsvTimeSeriesSource.java [hunk 2]
```diff
@@ -30,8 +66,181 @@ public CsvTimeSeriesSource(
...
+   * Reads in time series of a specified class from given {@link TimeSeriesReadingData} utilising a
+   * provided {@link TimeBasedSimpleValueFactory}.
+   *
+   * @param readingData Data needed for reading
+   * @param valueClass Class of the target value within the time series
+   * @param factory Factory to utilize
+   * @param <V> Type of the value
+   * @return A set of {@link IndividualTimeSeries}
+   */
+  private <V extends Value> Set<IndividualTimeSeries<V>> readIn(
```
Comment (johanneshiry) [inline src/main/java/edu/ie3/datamodel/io/source/csv/CsvTimeSeriesSource.java:153]:
```suggestion
  private <V extends Value> Set<IndividualTimeSeries<V>> read(
```

File: src/main/java/edu/ie3/datamodel/models/timeseries/TimeSeriesContainer.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+      Set<IndividualTimeSeries<HeatAndSValue>> heatAndApparentPower,
+      Set<IndividualTimeSeries<HeatAndPValue>> heatAndActivePower,
+      Set<IndividualTimeSeries<HeatDemandValue>> heat,
+      Set<IndividualTimeSeries<SValue>> apparentPower,
+      Set<IndividualTimeSeries<PValue>> activePower) {
+    this.weather = Objects.requireNonNull(weather, "Weather time series may not be null.");
+    this.energyPrice =
+        Objects.requireNonNull(energyPrice, "Energy price time series may not be null.");
+    this.heatAndApparentPower =
+        Objects.requireNonNull(
```
Comment (johanneshiry) [inline src/main/java/edu/ie3/datamodel/models/timeseries/TimeSeriesContainer.java:37]: 👍  for this approach!

File: src/main/java/edu/ie3/datamodel/io/source/csv/CsvRawGridSource.java [hunk 4]
```diff
@@ -468,8 +466,9 @@ public CsvRawGridSource(
    * @param <T> Type of the asset type
    * @return The enriched entity data
    */
-  private <T extends AssetTypeInput> TypedConnectorInputEntityData<T> addTypeToEntityData(
-      ConnectorInputEntityData untypedEntityData, T assetType) {
+  private <T extends AssetTypeInput, C extends ConnectorInput>
```
Comment (johanneshiry) [inline src/main/java/edu/ie3/datamodel/io/source/csv/CsvRawGridSource.java:469]:
```suggestion
  private <T extends AssetTypeInput>
```

**Review by johanneshiry — CHANGES_REQUESTED** [review]
Thank you very much for your work. I added some comments and change requests that I would prefer to be addressed before merging. In general I'm really looking forward having the time series I/O capabilities available on the PSDM soon! :)

File: src/main/java/edu/ie3/datamodel/io/factory/EntityFactory.java [hunk 5]
```diff
@@ -24,98 +19,21 @@
...
+    extends Factory<T, D, T> {
   /**
    * Constructor for an EntityFactory for given classes
    *
    * @param allowedClasses exactly the classes that this factory is allowed and able to build
    */
   public EntityFactory(Class<? extends T>... allowedClasses) {
-    this.classes = Arrays.asList(allowedClasses);
+    super(allowedClasses);
     TimeTools.initialize(ZoneId.of("UTC"), Locale.GERMANY, "yyyy-MM-dd HH:mm:ss");
```
[the remark this answers is not in the document]
Comment (ckittl) [reply src/main/java/edu/ie3/datamodel/io/factory/EntityFactory.java:31]: Intend to do this in a separate PR.

File: src/main/java/edu/ie3/datamodel/io/factory/result/ThermalResultFactory.java [hunk 6]
```diff
@@ -47,7 +46,7 @@ public ThermalResultFactory() {
 
   @Override
   protected ThermalUnitResult buildModel(SimpleEntityData data) {
-    Class<? extends UniqueEntity> clazz = data.getEntityClass();
+    Class<?> clazz = data.getTargetClass();
 
     ZonedDateTime zdtTimestamp = TimeTools.toZonedDateTime(data.getField(TIMESTAMP));
```
[the remark this answers is not in the document]
Comment (ckittl) [reply src/main/java/edu/ie3/datamodel/io/factory/result/ThermalResultFactory.java:51]: Intend to do this in a separate PR.

File: src/main/java/edu/ie3/datamodel/io/factory/timeseries/TimeBasedSimpleValueFactory.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    this(valueClasses, "yyyy-MM-dd'T'HH:mm:ss[.S[S][S]]'Z'");
+  }
+
+  public TimeBasedSimpleValueFactory(Class<? extends V> valueClasses, String timestampPattern) {
+    super(valueClasses);
+    timeUtil = new TimeUtil(ZoneId.of("UTC"), Locale.GERMANY, timestampPattern);
+  }
+
+  @Override
+  protected TimeBasedValue<V> buildModel(SimpleTimeBasedValueData<V> data) {
```
[the remark this answers is not in the document]
Comment (ckittl) [reply src/main/java/edu/ie3/datamodel/io/factory/timeseries/TimeBasedSimpleValueFactory.java:48]: At least we come to the same conclusion. 😂

**Comment by ckittl** [comment]
Removal of deprecated `TimeTools` will be addressed with #225.

**Comment by johanneshiry** [comment]
!test

**Review by johanneshiry — APPROVED** [review]

**Pull request merged**